clang/lib/Sema/SemaTemplate.cpp

```diff
@@ -2612,44 +2671,309 @@ struct ConvertConstructorToDeductionGuideTransform {
     SemaRef.CurrentInstantiationScope->InstantiatedLocal(OldParam, NewParam);
     return NewParam;
   }
+};
 
-  FunctionTemplateDecl *buildDeductionGuide(
-      TemplateParameterList *TemplateParams, CXXConstructorDecl *Ctor,
-      ExplicitSpecifier ES, TypeSourceInfo *TInfo, SourceLocation LocStart,
-      SourceLocation Loc, SourceLocation LocEnd,
-      llvm::ArrayRef<TypedefNameDecl *> MaterializedTypedefs = {}) {
-    DeclarationNameInfo Name(DeductionGuideName, Loc);
-    ArrayRef<ParmVarDecl *> Params =
-        TInfo->getTypeLoc().castAs<FunctionProtoTypeLoc>().getParams();
+// Find all template parameters of the AliasTemplate that appear in the
+// given DeducedArgs.
+SmallVector<unsigned>
+FindAppearedTemplateParamsInAlias(ArrayRef<TemplateArgument> DeducedArgs,
```

Comment:
@erichkeane

Reply:
I might suggest we instead genericize this to not be AliasTemplate specific, and have it take a list of NamedDecls and template args.  Then call it something like, `DeclsReferencedInTemplateArgumentList`.

I also probably wouldn't have this return a list of 'ints', and just return the pointers themselves, as this little trade back and forth is a little awkward.

Reply:
> I might suggest we instead genericize this to not be AliasTemplate specific, and have it take a list of NamedDecls and template args. Then call it something like, DeclsReferencedInTemplateArgumentList.

We could make this function more generic, but this function is only be used for alias-template CTAD implementation, and it is unlikely to be reused for other things, so the motivation to genericize it doesn't seem strong to me (happy to do this change if you feel strong about it).

> I also probably wouldn't have this return a list of 'ints', and just return the pointers themselves, as this little trade back and forth is a little awkward.

Yeah, I considered this in the initial version, but I found returning a list of Index can simplify the code. The caller of this function needs the `Index`, if we return a list of pointers, the caller needs to do some more work to figure out their corresponding Index.

Reply:
The generic suggestion is because it would, I believe, make this a much more readable function with very minor changes in the 'use'.  So I think there is value there.

I see what you mean about the index, I think it is alright to leave the index part.

Reply:
OK, that makes sense. I made it a generic function.